File: leetcode_java/java/com/goinhn/leetcode/_1_100/Sol21.java
/*
 * Copyright (C) 2021 Baidu, Inc. All Rights Reserved.
 */
package com.goinhn.leetcode._1_100;

/**
 * 21.合并两个有序链表
 *
 * @author fangwenzhan
 * @date 2021-11-05
 **/
public class Sol21 {

    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    /**
     * 顺序比较
     * 两个链表的长度分别为m、n
     * <p>
     * time:O(n + m)
     * space:O(1)
     */
    private static class Solution1 {
        public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
            ListNode tempHead = new ListNode(-1);
            ListNode resultHead = tempHead;

            if (null == l1 && null == l2) {
                return null;
            }
            if (l1 == null || l2 == null) {
                return l1 == null ? l2 : l1;
            }

            while (null != l1 && null != l2) {
                if (l1.val < l2.val) {
                    tempHead.next = l1;
                    l1 = l1.next;
                } else {
                    tempHead.next = l2;
                    l2 = l2.next;
                }
                tempHead = tempHead.next;
            }

            if (l1 != null) {
                tempHead.next = l1;
            }
            if (l2 != null) {
                tempHead.next = l2;
            }

            return resultHead.next;
        }
    }

    private static class Solution2 {
        public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
            if (null == l1) {
                return l2;
            } else if (null == l2) {
                return l1;
            } else if (l1.val < l2.val) {
                l1.next = mergeTwoLists(l1.next, l2);
                return l1;
            } else {
                l2.next = mergeTwoLists(l1, l2.next);
                return l2;
            }
        }
    }
}

File: leetcode_go/question/_1_100/_21/sol_21.go
// Package _21 合并两个有序链表
// @author goinhn
// @date 2022/8/16
package _21

type ListNode struct {
	Val  int
	Next *ListNode
}

// 顺序合并
// time:O(n)
// space:O(1)
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}

	if list2 == nil {
		return list1
	}

	preHead := &ListNode{Val: -1}
	pre := preHead
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			pre.Next = list1
			list1 = list1.Next
		} else {
			pre.Next = list2
			list2 = list2.Next
		}
		pre = pre.Next
	}

	if list1 != nil {
		pre.Next = list1
	}

	if list2 != nil {
		pre.Next = list2
	}

	return preHead.Next
}
